[mukeunzi] 로컬 로그인 구현

mukeunzi opened the pull request:
**passport local 로그인 구현**

- 회원가입
- 로그인
- 로그아웃

mukeunzi committed:
feat : express repo 생성

- express generator 사용
- .gitignore 생성

mukeunzi committed:
feat : README.md 파일 생성

mukeunzi committed:
feat : user 스키마 정의

mukeunzi committed:
feat : passport 패키지 설치

mukeunzi committed:
feat : npm 모듈 추가

mukeunzi committed:
feat : .env 파일 추가

mukeunzi committed:
feat : mongodb connection 코드 추가

mukeunzi committed:
feat : sign-up pug 추가

mukeunzi committed:
feat : sign-up submit 버튼 이벤트 추가

- submit 버튼이 클릭되면 기존 이벤트 무시하고 유효성검사 메소드 호출

mukeunzi committed:
feat : id/pwd 유효성 검사 메소드 추가

- id/pwd 가 비어있을 경우 리턴

mukeunzi committed:
feat : checkDuplicatedId 메소드 생성

- 회원가입 시 사용자가 입력한 아이디를 중복체크하는 메소드

mukeunzi committed:
feat : db connection 코드 수정

- db 연결 실패할 경우 프로세스 종료

mukeunzi committed:
refactor : userSchema statics 메소드 추가

- 회원가입 시 아이디를 중복검사하는 메소드 checkDuplicatedId() 추가

mukeunzi committed:
feat : GET /users 라우터 추가

- 회원가입 페이지

mukeunzi committed:
feat : POST /users API 추가

- 회원가입 라우터
- 역할 : 아이디 유효성검사, 회원가입

mukeunzi committed:
refactor : var 키워드 변경

- var => const 로 변경
- 회원가입 비밀번호 input name/id user_password 로 변경

mukeunzi committed:
feat : app.js 라우터 추가

- index, users 라우터 추가
- mongodb 연결 코드 추가

mukeunzi committed:
feat : auth 라우터 추가

- 로그인 페이지 GET /auth 추가

mukeunzi committed:
fix : addEventListener arrow function 수정

- addEventListener arrow function => 일반함수로 변경

mukeunzi committed:
feat : login html 추가

mukeunzi committed:
refactor : users form 유효성검사 util 로 변경

- 회원가입, 로그인에서 공통으로 사용하는 users-form util 로 추가

mukeunzi committed:
feat : passport local strategy 구현

- 로컬 passport 구현

mukeunzi committed:
feat : session, passport 미들웨어 추가

mukeunzi committed:
feat : passport index 추가

- serializeUser, deserializeUser 구현
- localStrategy 호출

mukeunzi committed:
feat : 로그인 인증 메소드 추가

- isAuthenticated() 메소드를 이용한 로그인/로그아웃 을 인증하는 메소드
생성

mukeunzi committed:
feat : 라우터에 login 인증 메소드 추가

mukeunzi committed:
feat : 로그인 api 추가

- localStrategy 사용

mukeunzi committed:
feat : 로그인시 비밀번호 암호화

- bcrypt 모듈 사용

mukeunzi committed:
fix : 로그인이 성공하면 '/' 로 리다이렉트

mukeunzi committed:
feat : 로그아웃 api 추가

- 로그인이 되어있을 경우 로그아웃을 실행하는 라우터

mukeunzi committed:
feat : flash 미들웨어 삭제

mukeunzi committed:
feat : flash 메시지 추가

- 로그인에 실패할 경우 flashMessage 출력

mukeunzi committed:
feat : npm 패키지 추가

- bcrypt, dotenv, express-session, mongoose
  , chai, mocha, sinon, supertest package install

mukeunzi committed:
refactor : deserializeUser 메소드 불필요한 코드 삭제

mukeunzi committed:
refactor : mongoose getUserInfo statics 메소드 생성

- user_id 로 회원 정보를 가져오는 메소드

mukeunzi committed:
refactor : signUp statics method 생성

- 회원가입을 하는 signUp 메소드
- 라우터에 있었던 mongoose 쿼리 모듈로 분리

mukeunzi committed:
refactor : isValidPassword methods 생성

- 로그인에서 입력한 비밀번호가 유효한지 체크하는 메소드

mukeunzi committed:
fix : clean-css package update version

mukeunzi committed:
fix : 스키마 메소드 호출 버그 수정

- [schema].method명 => [instance].method명으로 수정

mukeunzi committed:
feat : session을 이용한 로그인/로그아웃/회원가입 버튼 생성

- res.locals 변수에 req.user 할당
- index.pug 에서 user 가 있으면 로그아웃 버튼, 없으면 로그인/회원가입
버튼 생성

mukeunzi committed:
feat : 로그인/회원가입/로그아웃 버튼 클릭이벤트 추가

mukeunzi committed:
fix : users-form submit 버튼 이벤트 버그 수정

- isEmptyId(), isEmptyPassword() : 리턴 값 true/false 로 변경
- isValidFormData() : 유효성 검사 후 form 데이터 submit

mukeunzi committed:
feat : 로그아웃 버튼 이벤트 구현

- logout 버튼 => DELETE /auth 사용
- fetch api 사용으로 redirect 는 fetch response 후에 작동

mukeunzi committed:
feat : flash message 영역 추가

mukeunzi committed:
feat : 회원가입시 아이디 중복일 경우 flash message 사용

mukeunzi committed:
충돌 해결

honux77 commented on README.md:
```diff
@@ -31,7 +31,11 @@ Node.js 의 Express 프레임워크를 이용한 사이트입니다.
 **계정**
 
-- 사용자의 계정은 OAuth 방식만을 사용합니다.
+<<<<<<< HEAD
```
컥 충돌 해결 실패!

honux77 commented on models/index.js:
```diff
@@ -0,0 +1,35 @@
+		}
+
+		mongoose.connect(
+			`mongodb://${process.env.DB_USERNAME}:${process.env.DB_PASSWORD}@${process.env.DB_HOST}:${process.env.DB_PORT}/${
```
connect 메소드 중에 여러 매개변수를 받을 수 있는게 있는지 보고 있으면 그거 쓰는게 좋겠네요.

honux77 commented on models/index.js:
```diff
@@ -0,0 +1,35 @@
+
+	connect();
+
+	mongoose.connection.on('error', error => {
```
연결 오류가 발생할 경우 서비스가 죽어야 맞을 것 같네요. 실제 디비를 죽이고 잘 되는지 확인해 보세요.

honux77 commented on passport/index.js:
```diff
@@ -0,0 +1,13 @@
+const localStrategy = require('./local-strategy');
+
+module.exports = passport => {
```
시리얼에서 id를 던져주면 deserial에서 다시 디비에서 읽어오는 코드가 있어야 합니다!

honux77 commented on middlewares/login-auth.js:
```diff
@@ -0,0 +1,15 @@
+	res.status(403).send('로그인이 필요합니다.');
+};
+
+const isLoggedOut = (req, res, next) => {
```
이건 뭔지 모르겠네요? 로그인 상태가 아닌 경우에만 진행하는 것 같은데 그럼 isNotLoggedIn 이 좋을 것 같아요. 
그리고 로그인 상태에는 진입할 수 없는 페이지가 있다는 건 조금 이상하네요.
로그인 페이지가 그렇겠군요 ㅋㅋㅋ

mukeunzi replied:
isNotLoggedIn 으로 변경하겠습니다!
지금은 회원가입 페이지, 로그인 페이지에서 사용하고 있어요

honux77 reviewed (changes requested):
고칠게 조금 있어요.
확인해서 고치고 알려 주세요.

mukeunzi committed:
fix : README 충돌 해결

mukeunzi committed:
fix : deserializeUser 누락된 코드 수정

- user_id 로 회원 정보를 불러오는 코드 추가

mukeunzi committed:
refactor : isLoggedOut 메소드명 변경

- isLoggedOut => isNotLoggedIn 으로 변경
- 회원가입/로그인 페이지를 보여주기 전 로그인이 되어있지 않은 사용자인지
체크하는 메소드

mukeunzi committed:
refactor : mongodb connection 코드 수정

- (참고 : MongoDB 에서 deprecated 된 ensureIndex() warning 을 방지하기
위해 useCreateIndex() 추가)

mukeunzi committed:
Merge branch 'master' of https://github.com/mukeunzi/Vote-Your-Choice into mukeunzi

mukeunzi committed:
feat : 불필요한 주석 제거

mukeunzi committed:
Merge branch 'master' of https://github.com/mukeunzi/Vote-Your-Choice into mukeunzi

honux77 reviewed (approved):
승인!
